File: heuristic_library/cpp/SA.cpp
#include "template.cpp"

// パラメータ ///////////////////////////////////
// 提出用
constexpr float start_temp = 500;
constexpr float end_temp = 1;

// optuna 用
// constexpr float default_start_temp = 200;
// constexpr float default_end_temp = 1;
// float start_temp = default_start_temp;
// float end_temp = default_end_temp;
void get_param() {
    const char* p;
    // p = std::getenv("start_temp"); assert(p); start_temp = std::stof(p);
    // p = std::getenv("end_temp"); assert(p); end_temp = std::stof(p);
}
////////////////////////////////////////////////

// 焼きなましに関する関数 /////////////////////////
// 線形温度管理
float linear_temp(unsigned short &SA_start_time, unsigned short &now_time) {
    return start_temp - (start_temp - end_temp) * (now_time - SA_start_time) / time_limit;
}

// 指数温度管理
float expo_temp(unsigned short &SA_start_time, unsigned short &now_time) {
    float p = (float)(now_time - SA_start_time) / time_limit;
    return start_temp * pow(end_temp / start_temp, p);
}

// 遷移確率関数
constexpr bool maximize = false; // false: 最小化, true: 最大化

template<class Score>
bool is_better(const Score& a, const Score& b) {
    if constexpr (maximize) return a > b;
    else return a < b;
}

template<class Score>
double calc_prob(const Score& now_score, const Score& next_score, double temp) {
    if (is_better(next_score, now_score)) return 1.0;
    if constexpr (maximize) {
        return exp((next_score - now_score) / temp);
    } else {
        return exp((now_score - next_score) / temp);
    }
}

// スコアの型
using Score = int;
///////////////////////////////////////////////////////

// ここから下に解法を書く

// 焼きなましで使う構造体
// スコアの差分更新に使う配列などを持つ
struct WorkSpace {
    Score score;
};

// 最良解を保持するための構造体
// WorkSpace のうち、スコアと解の出力に必要な情報だけを持つ
struct Answer {
    Score score;
    // 「Answer = WorkSpace」という代入を可能にする (演算子オーバーロード)
    Answer& operator=(const WorkSpace& sol) {
        score = sol.score;
        return *this;
    }
};

// 入力

// 初期解生成
WorkSpace make_initial_solution(){
    WorkSpace res{};

    return res;
}

Score initialize_score(WorkSpace &sol) {
    Score score = 0;

    return score;
}

// 近傍生成 + スコア計算 + 受容判定 + 状態更新 -> 近傍生成が成功したかどうか を返す
// 状態は sol を直接書き換える
bool generate_neighborhood(auto &temp, WorkSpace &sol) {
    // 近傍生成


    // スコア計算 


    if (calc_prob(now_score, next_score, temp) > Random::random()) {

        return true;
    }
    else {
        // 状態をもとに戻す

        return true;
    }
}

Answer SA() {
    unsigned int counter = 0; unsigned int iter = 0; unsigned int failed_neighborhood = 0;
    auto SA_start_time = timer.get_ms();
    float temp = start_temp;

    WorkSpace current_solution = make_initial_solution();
    auto best_score = current_solution.score;
    Answer best_ans;
    best_ans = current_solution;
    cerr << "start score: " << current_solution.score << "\n";
    auto now_time = timer.get_ms();

    while (true) {
        if (counter == 30) {
            now_time = timer.get_ms();
            if (now_time > time_limit) break;
            temp = linear_temp(SA_start_time, now_time);
            counter = 0;
        }
        if (!generate_neighborhood(temp, current_solution)) {
            failed_neighborhood++;
        }
        if (is_better(current_solution.score, best_score)) {
            best_score = current_solution.score;
            best_ans = current_solution;
        }
        iter++; counter++;
    }
    cerr << "best score: " << best_score << "\n";
    cerr << "iter: " << iter << "\n";
    cerr << "successful neighborhood: " << iter - failed_neighborhood << "\n";
    return best_ans;
}

int main(){
    ios::sync_with_stdio(false); cin.tie(0);
    timer = Timer();
    // get_param(); // optuna を使うときはコメントアウトを外す

    // 入力 //////////////////////////////////////////
    
    //////////////////////////////////////////////////

    Answer best_ans = SA();

    // 出力 //////////////////////////////////////////

    //////////////////////////////////////////////////
}

File: heuristic_library/python/SA.py
import sys
def input():return sys.stdin.readline().rstrip()
def ii(): return int(input())
def ms(): return map(int, input().split())
def li(): return list(map(int,input().split()))
def print_err(*args): print(*args,file=sys.stderr)
from random import randrange,randint,random
from time import time
start = time()

from math import exp
# パラメータ ##############################
start_temp = 500
end_temp = 10
time_limit = 1.85 # 単位 : s
###########################################

# 線形温度管理
def linear_temp(SA_start_time : float, now_time : float) -> float:
  return start_temp - (start_temp - end_temp) * (now_time - SA_start_time) / time_limit

# 遷移確率関数
# スコア最大化のとき
def calc_prob_maximize(now_score, next_score, temp) -> float:
  return 1 if now_score < next_score else exp((now_score - next_score) / temp)

# スコア最小化のとき
def calc_prob_minimize(now_score, next_score, temp) -> float:
  return 1 if now_score > next_score else exp((next_score - now_score) / temp)

# グローバル変数 #########################

##########################################

def calc_score():

  return

def initialize_score():

  return

# 近傍生成 + スコア計算 + 受容判定 -> 新しいスコアを返す
def generate_neighbor(now_score : int|float, temp : float):
  # 近傍生成 ######################

  ################################
  # スコア計算 ###################
  next_score = 
  if (calc_prob_maximize(now_score, next_score, temp) > random()): # 受容する
    # 必要であれば状態を更新 ######
    
    ###############################
    return next_score
  else: # 受容しない
    # 状態を元に戻す #############

    ##############################
    return now_score
  
def SA():
  SA_start_time = time()
  iter = -1
  temp = start_temp
  # スコアの初期化
  now_score = initialize_score()
  while True:
    if (iter%500 == 0):
      now_time = time()
      if (now_time - start > time_limit):
        break
      temp = linear_temp(SA_start_time, now_time)
    now_score = generate_neighbor(now_score, temp)
    iter += 1
  print("SA DONE", iter, file=sys.stderr)

SA()

# 出力 ############################

##################################